File: src/main/java/io/github/cdgeass/formatter/visitor/CustomFromItemVisitor.java
package io.github.cdgeass.formatter.visitor;

import io.github.cdgeass.constants.StringConstants;
import net.sf.jsqlparser.schema.Table;
import net.sf.jsqlparser.statement.select.*;

/**
 * @author cdgeass
 * @since 2020-06-04
 */
public class CustomFromItemVisitor extends AbstractCustomVisitor implements FromItemVisitor {

    public CustomFromItemVisitor(int level) {
        super(level);
    }

    @Override
    public void visit(Table tableName) {
        append(tableName.toString());
    }

    @Override
    public void visit(SubSelect subSelect) {
        if (subSelect.isUseBrackets()) {
            append("(\n");
        }
        if (subSelect.getWithItemsList() != null && !subSelect.getWithItemsList().isEmpty()) {
            appendTab().append("WITH ");
            for (var iter = subSelect.getWithItemsList().iterator(); iter.hasNext(); ) {
                var withItem = iter.next();
                append(withItem.toString());
                if (iter.hasNext()) {
                    append(",");
                }
                append(" ");
            }
        }
        var customSelectVisitor = new CustomSelectVisitor(currentLevel());
        subSelect.getSelectBody().accept(customSelectVisitor);
        append(customSelectVisitor.toString());
        if (subSelect.isUseBrackets()) {
            append(StringConstants.LINE_BREAK).appendPreTab().append(")");
        }

        if (subSelect.getAlias() != null) {
            append(subSelect.getAlias().toString());
        }
        if (subSelect.getPivot() != null) {
            append(StringConstants.LINE_BREAK).appendTab().append(subSelect.getPivot().toString());
        }
        if (subSelect.getUnPivot() != null) {
            append(StringConstants.LINE_BREAK).appendTab().append(subSelect.getUnPivot().toString());
        }
    }

    @Override
    public void visit(SubJoin subjoin) {
        appendTab().append("(").append(subjoin.getLeft().toString());
        for (var join : subjoin.getJoinList()) {
            if (join.isSimple()) {
                append(", ").append(VisitorUtil.join(join, currentLevel()));
            } else {
                append(" ").append(VisitorUtil.join(join, currentLevel()));
            }
        }

        append(StringConstants.LINE_BREAK).appendTab().append(")");
        if (subjoin.getAlias() != null) {
            append(subjoin.getAlias().toString());
        }
        if (subjoin.getPivot() != null) {
            append(StringConstants.LINE_BREAK).appendTab().append(subjoin.getPivot().toString());
        }
        if (subjoin.getUnPivot() != null) {
            append(StringConstants.LINE_BREAK).appendTab().append(subjoin.getUnPivot().toString());
        }
    }

    @Override
    public void visit(LateralSubSelect lateralSubSelect) {
        var customFromItemVisitor = new CustomFromItemVisitor(currentLevel());
        lateralSubSelect.accept(customFromItemVisitor);
        append(customFromItemVisitor.toString());
    }

    @Override
    public void visit(ValuesList valuesList) {
        appendTab().append("(VALUES ");
        for (var iter = valuesList.getMultiExpressionList().getExprList().iterator(); iter.hasNext(); ) {
            append(VisitorUtil.getStringList(iter.next().getExpressions(), true, !valuesList.isNoBrackets()));
            if (iter.hasNext()) {
                append(", ");
            }
        }
        append(StringConstants.LINE_BREAK).appendTab().append(")");
        if (valuesList.getAlias() != null) {
            append(valuesList.getAlias().toString());

            if (valuesList.getColumnNames() != null) {
                append(StringConstants.LINE_BREAK).appendTab().append("(");
                for (var iter = valuesList.getColumnNames().iterator(); iter.hasNext(); ) {
                    append(iter.next());
                    if (iter.hasNext()) {
                        append(", ");
                    }
                }
                append(StringConstants.LINE_BREAK).appendTab().append(")");
            }
        }
    }

    @Override
    public void visit(TableFunction tableFunction) {
        var customFromItemVisitor = new CustomFromItemVisitor(currentLevel());
        tableFunction.accept(customFromItemVisitor);
        append(customFromItemVisitor.toString());
    }

    @Override
    public void visit(ParenthesisFromItem aThis) {
        appendTab().append("(");
        var customFromItemVisitor = new CustomFromItemVisitor(nextLevel());
        aThis.getFromItem().accept(customFromItemVisitor);
        append(customFromItemVisitor.toString());
        append(StringConstants.LINE_BREAK).appendTab().append(")");

        if (aThis.getAlias() != null) {
            append(aThis.getAlias().toString());
        }
    }
}

File: src/main/kotlin/io/github/cdgeass/formatter/visitor/CustomFromItemVisitor.kt
package io.github.cdgeass.formatter.visitor

import net.sf.jsqlparser.schema.Table
import net.sf.jsqlparser.statement.select.*

/**
 * @author cdgeass
 * @since 2021-03-05
 */
class CustomFromItemVisitor(level: Int) : AbstractCustomVisitor(level), FromItemVisitor {

    override fun visit(tableName: Table?) {
        appendTab().append(tableName?.toString() ?: "")
    }

    override fun visit(subSelect: SubSelect?) {
        if (subSelect?.isUseBrackets == true) {
            append("(")
        }
        if (subSelect?.withItemsList?.isNotEmpty() == true) {
            appendTab().append("WITH ").append(subSelect.withItemsList.joinToString(",") { toString() })
        }

        append(CustomSelectVisitor(currentLevel()).apply { subSelect?.selectBody?.accept(this) }.toString())
        if (subSelect?.isUseBrackets == true) {
            appendPreTab().append(")")
        }

        if (subSelect?.alias != null) {
            append(subSelect.alias.toString())
        }
        if (subSelect?.pivot != null) {
            appendTab().append(subSelect.pivot.toString())
        }
        if (subSelect?.unPivot != null) {
            appendTab().append(subSelect.unPivot.toString())
        }
    }

    override fun visit(subjoin: SubJoin?) {
        appendTab().append("(").append(subjoin?.left?.toString() ?: "")
        subjoin?.joinList?.forEach {
            if (it.isSimple) {
                append(", ").append(join(it, currentLevel()))
            } else {
                append(" ").append(join(it, currentLevel()))
            }
        }
        appendTab().append(")")

        if (subjoin?.alias != null) {
            append(subjoin.alias.toString())
        }
        if (subjoin?.pivot != null) {
            append(subjoin.pivot.toString())
        }
        if (subjoin?.unPivot != null) {
            append(subjoin.unPivot.toString())
        }
    }

    override fun visit(lateralSubSelect: LateralSubSelect?) {
        appendTab().append(CustomFromItemVisitor(currentLevel()).apply { lateralSubSelect?.accept(this) }.toString())
    }

    override fun visit(valuesList: ValuesList?) {
        appendTab()
            .append("(VALUES ")
            .append(
                valuesList?.multiExpressionList?.expressionLists?.joinToString(",") { exprList ->
                    getStringList(exprList.expressions, true, !valuesList.isNoBrackets)
                } ?: ""
            )
            .append(")")

        if (valuesList?.alias != null) {
            append(valuesList.alias.toString())

            if (valuesList.columnNames?.isNotEmpty() == true) {
                appendTab().append("(").append(valuesList.columnNames.joinToString(",")).append(")")
            }
        }
    }

    override fun visit(tableFunction: TableFunction?) {
        appendTab().append(CustomFromItemVisitor(currentLevel()).apply { tableFunction?.accept(this) }.toString())
    }

    override fun visit(aThis: ParenthesisFromItem?) {
        appendTab()
            .append("(")
            .append(CustomFromItemVisitor(currentLevel()).apply { aThis?.fromItem?.accept(this) }.toString())
            .append(")")

        if (aThis?.alias != null) {
            append(aThis.alias.toString())
        }
    }
}